DensityAnalysis Normalization

The current density function may have an issue with normalization. Since DensityAnalysis uses the MDAnalysis density function, it is currently unclear whether MDAnalysis.density.density_from_universe() is incorrect or the implementation of density_from_universe() into pmda is incorrect. We need to compare the results from a run of MDAnalysis.analysis.density.density_from_universe() to pmda.density.DensityAnalysis for other trajectories to see if they agree, as well as find a 3rd party method of calculating densities to compare values with.

Can you please also look into why DensityAnalysis failed after merging into master? See https://travis-ci.org/MDAnalysis/pmda/jobs/592141990
```
=================================== FAILURES ===================================

_____________________________ test_n_frames[0-5-2] _____________________________

u = <Universe with 43480 atoms>, start = 0, stop = 5, step = 2

    @pytest.mark.parametrize("step", [1, 2])

    @pytest.mark.parametrize("stop", [5, 4])

    @pytest.mark.parametrize("start", [0, 1])

    def test_n_frames(u, start, stop, step):

        D = pmda.density.DensityAnalysis(u.atoms)

        D.run(start, stop, step)

>       assert D._n_frames == len(range(start, stop, step))

E       assert 5 == 3

E        +  where 5 = <pmda.density.DensityAnalysis object at 0x7f1d0ac42e48>._n_frames

E        +  and   3 = len(range(0, 5, 2))

E        +    where range(0, 5, 2) = range(0, 5, 2)

/home/travis/build/MDAnalysis/pmda/pmda/test/test_density.py:85: AssertionError

_____________________________ test_n_frames[0-4-1] _____________________________

u = <Universe with 43480 atoms>, start = 0, stop = 4, step = 1

    @pytest.mark.parametrize("step", [1, 2])

    @pytest.mark.parametrize("stop", [5, 4])

    @pytest.mark.parametrize("start", [0, 1])

    def test_n_frames(u, start, stop, step):

        D = pmda.density.DensityAnalysis(u.atoms)

        D.run(start, stop, step)

>       assert D._n_frames == len(range(start, stop, step))

E       assert 5 == 4

E        +  where 5 = <pmda.density.DensityAnalysis object at 0x7f1d4494a160>._n_frames

E        +  and   4 = len(range(0, 4))

E        +    where range(0, 4) = range(0, 4, 1)

/home/travis/build/MDAnalysis/pmda/pmda/test/test_density.py:85: AssertionError

_____________________________ test_n_frames[0-4-2] _____________________________

u = <Universe with 43480 atoms>, start = 0, stop = 4, step = 2

    @pytest.mark.parametrize("step", [1, 2])

    @pytest.mark.parametrize("stop", [5, 4])

    @pytest.mark.parametrize("start", [0, 1])

    def test_n_frames(u, start, stop, step):

        D = pmda.density.DensityAnalysis(u.atoms)

        D.run(start, stop, step)

>       assert D._n_frames == len(range(start, stop, step))

E       assert 5 == 2

E        +  where 5 = <pmda.density.DensityAnalysis object at 0x7f1d2f8eca90>._n_frames

E        +  and   2 = len(range(0, 4, 2))

E        +    where range(0, 4, 2) = range(0, 4, 2)

/home/travis/build/MDAnalysis/pmda/pmda/test/test_density.py:85: AssertionError

_____________________________ test_n_frames[1-5-1] _____________________________

u = <Universe with 43480 atoms>, start = 1, stop = 5, step = 1

    @pytest.mark.parametrize("step", [1, 2])

    @pytest.mark.parametrize("stop", [5, 4])

    @pytest.mark.parametrize("start", [0, 1])

    def test_n_frames(u, start, stop, step):

        D = pmda.density.DensityAnalysis(u.atoms)

        D.run(start, stop, step)

>       assert D._n_frames == len(range(start, stop, step))

E       assert 5 == 4

E        +  where 5 = <pmda.density.DensityAnalysis object at 0x7f1d1ba772b0>._n_frames

E        +  and   4 = len(range(1, 5))

E        +    where range(1, 5) = range(1, 5, 1)

/home/travis/build/MDAnalysis/pmda/pmda/test/test_density.py:85: AssertionError

_____________________________ test_n_frames[1-5-2] _____________________________

u = <Universe with 43480 atoms>, start = 1, stop = 5, step = 2

    @pytest.mark.parametrize("step", [1, 2])

    @pytest.mark.parametrize("stop", [5, 4])

    @pytest.mark.parametrize("start", [0, 1])

    def test_n_frames(u, start, stop, step):

        D = pmda.density.DensityAnalysis(u.atoms)

        D.run(start, stop, step)

>       assert D._n_frames == len(range(start, stop, step))

E       assert 5 == 2

E        +  where 5 = <pmda.density.DensityAnalysis object at 0x7f1d4f8e1940>._n_frames

E        +  and   2 = len(range(1, 5, 2))

E        +    where range(1, 5, 2) = range(1, 5, 2)

/home/travis/build/MDAnalysis/pmda/pmda/test/test_density.py:85: AssertionError

_____________________________ test_n_frames[1-4-1] _____________________________

u = <Universe with 43480 atoms>, start = 1, stop = 4, step = 1

    @pytest.mark.parametrize("step", [1, 2])

    @pytest.mark.parametrize("stop", [5, 4])

    @pytest.mark.parametrize("start", [0, 1])

    def test_n_frames(u, start, stop, step):

        D = pmda.density.DensityAnalysis(u.atoms)

        D.run(start, stop, step)

>       assert D._n_frames == len(range(start, stop, step))

E       assert 5 == 3

E        +  where 5 = <pmda.density.DensityAnalysis object at 0x7f1d090fd6a0>._n_frames

E        +  and   3 = len(range(1, 4))

E        +    where range(1, 4) = range(1, 4, 1)

/home/travis/build/MDAnalysis/pmda/pmda/test/test_density.py:85: AssertionError

_____________________________ test_n_frames[1-4-2] _____________________________

u = <Universe with 43480 atoms>, start = 1, stop = 4, step = 2

    @pytest.mark.parametrize("step", [1, 2])

    @pytest.mark.parametrize("stop", [5, 4])

    @pytest.mark.parametrize("start", [0, 1])

    def test_n_frames(u, start, stop, step):

        D = pmda.density.DensityAnalysis(u.atoms)

        D.run(start, stop, step)

>       assert D._n_frames == len(range(start, stop, step))

E       assert 5 == 2

E        +  where 5 = <pmda.density.DensityAnalysis object at 0x7f1d1bb6f9e8>._n_frames

E        +  and   2 = len(range(1, 4, 2))

E        +    where range(1, 4, 2) = range(1, 4, 2)

/home/travis/build/MDAnalysis/pmda/pmda/test/test_density.py:85: AssertionError
```

--- a/pmda/density.py
+++ b/pmda/density.py
@@ -304,7 +304,7 @@ class DensityAnalysis(ParallelAnalysisBase):
         metadata['psf'] = self._atomgroup.universe.filename
         metadata['dcd'] = self._trajectory.filename
         metadata['atomselection'] = self._atomselection
-        # metadata['n_frames'] = self._n_frames
+        metadata['n_frames'] = self.n_frames
         metadata['totaltime'] = self._atomgroup.universe.trajectory.totaltime
         metadata['dt'] = self._trajectory.dt
         metadata['time_unit'] = mda.core.flags['time_unit']

--- a/pmda/density.py
+++ b/pmda/density.py
@@ -251,7 +251,6 @@ class DensityAnalysis(ParallelAnalysisBase):
         self._ydim = ydim
         self._zdim = zdim
         self._trajectory = u.trajectory
-        self._n_frames = u.trajectory.n_frames
         if updating and atomselection is None:
             raise ValueError("updating=True requires a atomselection string")
         elif not updating and atomselection is not None:
@@ -300,12 +299,12 @@ class DensityAnalysis(ParallelAnalysisBase):
 
     def _conclude(self):
         self._grid = self._results[:].sum(axis=0)
-        self._grid /= float(self._n_frames)
+        self._grid /= float(self.n_frames)
         metadata = self._metadata if self._metadata is not None else {}
         metadata['psf'] = self._atomgroup.universe.filename
         metadata['dcd'] = self._trajectory.filename
         metadata['atomselection'] = self._atomselection
-        metadata['n_frames'] = self._n_frames
+        # metadata['n_frames'] = self._n_frames
         metadata['totaltime'] = self._atomgroup.universe.trajectory.totaltime
         metadata['dt'] = self._trajectory.dt
         metadata['time_unit'] = mda.core.flags['time_unit']

--- a/pmda/parallel.py
+++ b/pmda/parallel.py
@@ -352,6 +352,8 @@ class ParallelAnalysisBase(object):
                                                                  stop, step)
         n_frames = len(range(start, stop, step))
 
+        self.n_frames = n_frames
+
         if n_frames == 0:
             warnings.warn("run() analyses no frames: check start/stop/step")
         if n_frames < n_blocks:

--- a/pmda/rms/rmsf.py
+++ b/pmda/rms/rmsf.py
@@ -193,10 +193,9 @@ class RMSF(ParallelAnalysisBase):
         # serial case
         if n_blocks == 1:
             # get length of trajectory slice
-            k = len(self._blocks[0])
             self.mean = self._results[0, 0]
             self.sumsquares = self._results[0, 1]
-            self.rmsf = np.sqrt(self.sumsquares.sum(axis=1) / k)
+            self.rmsf = np.sqrt(self.sumsquares.sum(axis=1) / self.n_frames)
         # parallel case
         else:
             mean = self._results[:, 0]
@@ -207,10 +206,9 @@ class RMSF(ParallelAnalysisBase):
                 vals.append((len(self._blocks[i]), mean[i], sos[i]))
             # combine block results using fold method
             results = fold_second_order_moments(vals)
-            self.totalts = results[0]
             self.mean = results[1]
             self.sumsquares = results[2]
-            self.rmsf = np.sqrt(self.sumsquares.sum(axis=1) / self.totalts)
+            self.rmsf = np.sqrt(self.sumsquares.sum(axis=1) / self.n_frames)
             self._negative_rmsf(self.rmsf)
 
     @staticmethod

--- a/pmda/test/test_density.py
+++ b/pmda/test/test_density.py
@@ -1,76 +1,77 @@
 import pytest
 import MDAnalysis as mda
 import numpy as np
-import pmda.density
+from pmda.density import DensityAnalysis
 from MDAnalysis.analysis.density import density_from_Universe
 from numpy.testing import assert_almost_equal
 from MDAnalysisTests.datafiles import GRO_MEMPROT, XTC_MEMPROT
+from MDAnalysisTests.datafiles import PSF_TRICLINIC, DCD_TRICLINIC
 
 
 @pytest.fixture(scope='module')
-def u():
+def u1():
     return mda.Universe(GRO_MEMPROT, XTC_MEMPROT)
 
 
-def test_density_sum(u):
-    pdensity = pmda.density.DensityAnalysis(u.atoms, atomselection='name OD1',
-                                            updating=True)
-    core_number = 2
-    pdensity.run(n_blocks=core_number, n_jobs=core_number)
-    dens = mda.analysis.density.density_from_Universe(u,
-                                                      atomselection='name OD1',
-                                                      update_selection=True)
-    assert np.sum(dens.grid) == np.sum(pdensity.density.grid)
+@pytest.fixture(scope='module')
+def u2():
+    return mda.Universe(PSF_TRICLINIC, DCD_TRICLINIC)
 
 
-@pytest.mark.parametrize("n_blocks", [1, 2, 3, 4])
-def test_density_grid(u, n_blocks):
-    pdensity = pmda.density.DensityAnalysis(u.atoms, atomselection='name OD1',
-                                            updating=True)
-    core_number = n_blocks
-    pdensity.run(n_blocks=core_number, n_jobs=core_number)
-    dens = mda.analysis.density.density_from_Universe(u,
-                                                      atomselection='name OD1',
-                                                      update_selection=True)
-    assert_almost_equal(dens.grid, pdensity.density.grid)
+@pytest.mark.parametrize("n_blocks", [1, 2])
+@pytest.mark.parametrize("stop", [4, 5])
+@pytest.mark.parametrize("step", [1, 2])
+def test_density_values(u1, n_blocks, stop, step):
+    parallel = DensityAnalysis(u1.atoms, atomselection='name OD1',
+                               updating=True)
+    parallel.run(n_blocks=n_blocks, n_jobs=n_blocks, start=0, stop=stop,
+                 step=step)
+    serial = density_from_Universe(u1, atomselection='name OD1',
+                                   update_selection=True, start=0, stop=stop,
+                                   step=step)
+    assert np.sum(serial.grid) == np.sum(parallel.density.grid)
+    assert_almost_equal(serial.grid, parallel.density.grid, decimal=17)
 
 
-def test_updating(u):
+def test_updating(u1):
     with pytest.raises(ValueError):
-        pdensity = pmda.density.DensityAnalysis(u.atoms,
-                                                updating=True)
+        pdensity = DensityAnalysis(u1.atoms, updating=True)
 
 
-def test_atomselection(u):
+def test_atomselection(u1):
     with pytest.raises(ValueError):
-        pdensity = pmda.density.DensityAnalysis(u.atoms,
-                                                atomselection='name OD1')
+        pdensity = DensityAnalysis(u1.atoms, atomselection='name OD1')
 
 
-def test_gridcenter(u):
-    aselect = 'name OD1'
+def test_gridcenter(u1):
     gridcenter = np.array([10, 10, 10])
     xdim = 190
     ydim = 200
     zdim = 210
-    dens = mda.analysis.density.density_from_Universe(u,
-                                                      atomselection=aselect,
-                                                      update_selection=True,
-                                                      gridcenter=gridcenter,
-                                                      xdim=xdim,
-                                                      ydim=ydim,
-                                                      zdim=zdim)
-    pdens = pmda.density.DensityAnalysis(u.atoms,
-                                         atomselection=aselect,
-                                         updating=True,
-                                         gridcenter=gridcenter,
-                                         xdim=xdim,
-                                         ydim=ydim,
-                                         zdim=zdim)
+    serial = density_from_Universe(u1, atomselection='name OD1',
+                                   update_selection=True,
+                                   gridcenter=gridcenter,
+                                   xdim=xdim,
+                                   ydim=ydim,
+                                   zdim=zdim)
+    parallel = DensityAnalysis(u1.atoms, atomselection='name OD1',
+                               updating=True,
+                               gridcenter=gridcenter,
+                               xdim=xdim,
+                               ydim=ydim,
+                               zdim=zdim)
     core_number = 4
-    pdens.run(n_blocks=core_number, n_jobs=core_number)
-    assert_almost_equal(dens.grid, pdens.density.grid)
-    assert_almost_equal(pdens._gridcenter, gridcenter)
-    assert len(pdens.density.edges[0]) == xdim + 1
-    assert len(pdens.density.edges[1]) == ydim + 1
-    assert len(pdens.density.edges[2]) == zdim + 1
+    parallel.run(n_blocks=core_number, n_jobs=core_number)
+    assert_almost_equal(serial.grid, parallel.density.grid)
+    assert_almost_equal(parallel._gridcenter, gridcenter)
+    assert len(parallel.density.edges[0]) == xdim + 1
+    assert len(parallel.density.edges[1]) == ydim + 1
+    assert len(parallel.density.edges[2]) == zdim + 1
+
+
+@pytest.mark.parametrize("start", [0, 1])
+@pytest.mark.parametrize("stop", [5, 7, 10])
+@pytest.mark.parametrize("step", [1, 2, 3])
+def test_n_frames(u2, start, stop, step):
+    pdensity = DensityAnalysis(u2.atoms).run(start, stop, step)
+    assert pdensity.n_frames == len(range(start, stop, step))

--- a/pmda/test/test_rmsf.py
+++ b/pmda/test/test_rmsf.py
@@ -32,7 +32,8 @@ def test_rmsf_accuracy(u, n_blocks, decimal, step):
                                   n_jobs=n_blocks)
     MDA_vals = mda.analysis.rms.RMSF(u.atoms).run(step=step)
     assert_almost_equal(MDA_vals.mean, PMDA_vals.mean, decimal=decimal)
-    assert_almost_equal(MDA_vals.sumsquares, PMDA_vals.sumsquares, decimal=decimal)
+    assert_almost_equal(MDA_vals.sumsquares, PMDA_vals.sumsquares,
+                        decimal=decimal)
     assert_almost_equal(MDA_vals.rmsf, PMDA_vals.rmsf, decimal=decimal)
 
 

--- a/pmda/test/test_rmsf.py
+++ b/pmda/test/test_rmsf.py
@@ -23,8 +23,21 @@ def u():
     return mda.Universe(PSF, DCD)
 
 
+@pytest.mark.parametrize('n_blocks', (2, 3, 4))
+@pytest.mark.parametrize('decimal', (7, 10, 11))
+@pytest.mark.parametrize("step", [1, 2, 3, 5])
+def test_rmsf_accuracy(u, n_blocks, decimal, step):
+    PMDA_vals = RMSF(u.atoms).run(step=step,
+                                  n_blocks=n_blocks,
+                                  n_jobs=n_blocks)
+    MDA_vals = mda.analysis.rms.RMSF(u.atoms).run(step=step)
+    assert_almost_equal(MDA_vals.mean, PMDA_vals.mean, decimal=decimal)
+    assert_almost_equal(MDA_vals.sumsquares, PMDA_vals.sumsquares, decimal=decimal)
+    assert_almost_equal(MDA_vals.rmsf, PMDA_vals.rmsf, decimal=decimal)
+
+
 @pytest.mark.parametrize('n_blocks', (1, 2, 3, 4, 5, 10))
-@pytest.mark.parametrize('n_frames', (10, 50, 100))
+@pytest.mark.parametrize('n_frames', (10, 45, 90))
 def test_RMSF_values(u, n_blocks, n_frames):
     PMDA_vals = RMSF(u.atoms).run(stop=n_frames,
                                   n_blocks=n_blocks,
@@ -49,3 +62,11 @@ def test_RMSF_n_jobs(u, n_blocks):
 def test_negative_RMSF_raises_ValueError():
     with pytest.raises(ValueError):
         RMSF._negative_rmsf(np.array([-1, -1, -1]))
+
+
+@pytest.mark.parametrize("stop", [10, 33, 45, 90])
+@pytest.mark.parametrize("step", [1, 2, 3, 5])
+def test_n_frames(u, stop, step):
+    F = RMSF(u.atoms)
+    F.run(stop=stop, step=step)
+    assert F.n_frames == len(range(0, stop, step))

